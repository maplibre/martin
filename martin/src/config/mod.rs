pub mod args;
pub mod file;
