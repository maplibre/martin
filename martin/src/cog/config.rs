use std::fmt::Debug;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use url::Url;

use super::source::CogSource;
use crate::Source;
use crate::config::UnrecognizedValues;
use crate::file_config::{ConfigExtras, FileResult, SourceConfigExtras};

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct CogConfig {
    #[serde(flatten)]
    pub unrecognized: UnrecognizedValues,
}

impl ConfigExtras for CogConfig {
    fn get_unrecognized(&self) -> &UnrecognizedValues {
        &self.unrecognized
    }
}

impl SourceConfigExtras for CogConfig {
    async fn new_sources(&self, id: String, path: PathBuf) -> FileResult<Box<dyn Source>> {
        let cog = CogSource::new(id, path)?;
        Ok(Box::new(cog))
    }

    #[allow(clippy::no_effect_underscore_binding)]
    async fn new_sources_url(&self, _id: String, _url: Url) -> FileResult<Box<dyn Source>> {
        unreachable!()
    }

    fn parse_urls() -> bool {
        false
    }
}
