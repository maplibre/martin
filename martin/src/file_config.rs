use std::collections::{BTreeMap, HashSet};
use std::fmt::Debug;
use std::mem;
use std::path::{Path, PathBuf};

use futures::TryFutureExt;
use log::{info, warn};
use serde::{Deserialize, Serialize};
use url::Url;

use crate::MartinResult;
use crate::OptOneMany::{Many, One};
use crate::config::{UnrecognizedValues, copy_unrecognized_config};
use crate::file_config::FileError::{
    InvalidFilePath, InvalidSourceFilePath, InvalidSourceUrl, IoError,
};
use crate::source::{TileInfoSource, TileInfoSources};
use crate::utils::{IdResolver, OptMainCache, OptOneMany};

pub type FileResult<T> = Result<T, FileError>;

#[derive(thiserror::Error, Debug)]
pub enum FileError {
    #[error("IO error {0}: {1}")]
    IoError(std::io::Error, PathBuf),

    #[error("Source path is not a file: {0}")]
    InvalidFilePath(PathBuf),

    #[error("Error {0} while parsing URL {1}")]
    InvalidSourceUrl(url::ParseError, String),

    #[error("Source {0} uses bad file {1}")]
    InvalidSourceFilePath(String, PathBuf),

    #[error(r"Unable to parse metadata in file {1}: {0}")]
    InvalidMetadata(String, PathBuf),

    #[error(r"Unable to parse metadata in file {1}: {0}")]
    InvalidUrlMetadata(String, Url),

    #[error(r"Unable to acquire connection to file: {0}")]
    AcquireConnError(String),

    #[cfg(feature = "pmtiles")]
    #[error(r"PMTiles error {0} processing {1}")]
    PmtError(pmtiles::PmtError, String),

    #[cfg(feature = "cog")]
    #[error(transparent)]
    CogError(#[from] crate::cog::CogError),
}

pub trait ConfigExtras: Clone + Debug + Default + PartialEq + Send {
    fn init_parsing(&mut self, _cache: OptMainCache) -> FileResult<()> {
        Ok(())
    }

    #[must_use]
    fn is_default(&self) -> bool {
        true
    }

    fn get_unrecognized(&self) -> &UnrecognizedValues;
}

pub trait SourceConfigExtras: ConfigExtras {
    #[must_use]
    fn parse_urls() -> bool {
        false
    }

    fn new_sources(
        &self,
        id: String,
        path: PathBuf,
    ) -> impl Future<Output = FileResult<TileInfoSource>> + Send;

    fn new_sources_url(
        &self,
        id: String,
        url: Url,
    ) -> impl Future<Output = FileResult<TileInfoSource>> + Send;
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum FileConfigEnum<T> {
    #[default]
    None,
    Path(PathBuf),
    Paths(Vec<PathBuf>),
    Config(FileConfig<T>),
}

impl<T: ConfigExtras> FileConfigEnum<T> {
    #[must_use]
    pub fn new(paths: Vec<PathBuf>) -> FileConfigEnum<T> {
        Self::new_extended(paths, BTreeMap::new(), T::default())
    }

    #[must_use]
    pub fn new_extended(
        paths: Vec<PathBuf>,
        configs: BTreeMap<String, FileConfigSrc>,
        custom: T,
    ) -> Self {
        if configs.is_empty() && custom.is_default() {
            match paths.len() {
                0 => FileConfigEnum::None,
                1 => FileConfigEnum::Path(paths.into_iter().next().unwrap()),
                _ => FileConfigEnum::Paths(paths),
            }
        } else {
            FileConfigEnum::Config(FileConfig {
                paths: OptOneMany::new(paths),
                sources: if configs.is_empty() {
                    None
                } else {
                    Some(configs)
                },
                custom,
            })
        }
    }

    #[must_use]
    pub fn is_none(&self) -> bool {
        matches!(self, Self::None)
    }

    #[must_use]
    pub fn is_empty(&self) -> bool {
        match self {
            Self::None => true,
            Self::Path(_) => false,
            Self::Paths(v) => v.is_empty(),
            Self::Config(c) => c.is_empty(),
        }
    }

    pub fn extract_file_config(
        &mut self,
        cache: OptMainCache,
    ) -> FileResult<Option<FileConfig<T>>> {
        let mut res = match self {
            FileConfigEnum::None => return Ok(None),
            FileConfigEnum::Path(path) => FileConfig {
                paths: One(mem::take(path)),
                ..FileConfig::default()
            },
            FileConfigEnum::Paths(paths) => FileConfig {
                paths: Many(mem::take(paths)),
                ..Default::default()
            },
            FileConfigEnum::Config(cfg) => mem::take(cfg),
        };
        res.custom.init_parsing(cache)?;
        Ok(Some(res))
    }

    pub fn finalize(&self, prefix: &str) -> UnrecognizedValues {
        let mut res = UnrecognizedValues::new();
        if let Self::Config(cfg) = self {
            copy_unrecognized_config(&mut res, prefix, cfg.get_unrecognized());
        }
        res
    }
}

#[serde_with::skip_serializing_none]
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct FileConfig<T> {
    /// A list of file paths
    #[serde(default, skip_serializing_if = "OptOneMany::is_none")]
    pub paths: OptOneMany<PathBuf>,
    /// A map of source IDs to file paths or config objects
    pub sources: Option<BTreeMap<String, FileConfigSrc>>,
    /// Any customizations related to the specifics of the configuration section
    #[serde(flatten)]
    pub custom: T,
}

impl<T: ConfigExtras> FileConfig<T> {
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.paths.is_none()
            && self.sources.is_none()
            && self.get_unrecognized().is_empty()
            && self.custom.is_default()
    }

    pub fn get_unrecognized(&self) -> &UnrecognizedValues {
        self.custom.get_unrecognized()
    }
}

/// A serde helper to store a boolean as an object.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum FileConfigSrc {
    Path(PathBuf),
    Obj(FileConfigSource),
}

impl FileConfigSrc {
    #[must_use]
    pub fn into_path(self) -> PathBuf {
        match self {
            Self::Path(p) => p,
            Self::Obj(o) => o.path,
        }
    }

    #[must_use]
    pub fn get_path(&self) -> &PathBuf {
        match self {
            Self::Path(p) => p,
            Self::Obj(o) => &o.path,
        }
    }

    pub fn abs_path(&self) -> FileResult<PathBuf> {
        let path = self.get_path();
        path.canonicalize().map_err(|e| IoError(e, path.clone()))
    }
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct FileConfigSource {
    pub path: PathBuf,
}

pub async fn resolve_files<T: SourceConfigExtras>(
    config: &mut FileConfigEnum<T>,
    idr: &IdResolver,
    cache: OptMainCache,
    extension: &[&str],
) -> MartinResult<TileInfoSources> {
    resolve_int(config, idr, cache, extension)
        .map_err(crate::MartinError::from)
        .await
}

async fn resolve_int<T: SourceConfigExtras>(
    config: &mut FileConfigEnum<T>,
    idr: &IdResolver,
    cache: OptMainCache,
    extension: &[&str],
) -> FileResult<TileInfoSources> {
    let Some(cfg) = config.extract_file_config(cache)? else {
        return Ok(TileInfoSources::default());
    };

    let mut results = TileInfoSources::default();
    let mut configs = BTreeMap::new();
    let mut files = HashSet::new();
    let mut directories = Vec::new();

    if let Some(sources) = cfg.sources {
        for (id, source) in sources {
            if let Some(url) = parse_url(T::parse_urls(), source.get_path())? {
                let dup = !files.insert(source.get_path().clone());
                let dup = if dup { "duplicate " } else { "" };
                let id = idr.resolve(&id, url.to_string());
                configs.insert(id.clone(), source);
                results.push(cfg.custom.new_sources_url(id.clone(), url.clone()).await?);
                info!("Configured {dup}source {id} from {}", sanitize_url(&url));
            } else {
                let can = source.abs_path()?;
                if !can.is_file() {
                    // todo: maybe warn instead?
                    return Err(InvalidSourceFilePath(id.to_string(), can));
                }

                let dup = !files.insert(can.clone());
                let dup = if dup { "duplicate " } else { "" };
                let id = idr.resolve(&id, can.to_string_lossy().to_string());
                info!("Configured {dup}source {id} from {}", can.display());
                configs.insert(id.clone(), source.clone());
                results.push(cfg.custom.new_sources(id, source.into_path()).await?);
            }
        }
    }

    for path in cfg.paths {
        if let Some(url) = parse_url(T::parse_urls(), &path)? {
            let target_ext = extension.iter().find(|&e| url.to_string().ends_with(e));
            let id = if let Some(ext) = target_ext {
                url.path_segments()
                    .and_then(Iterator::last)
                    .and_then(|s| {
                        // Strip extension and trailing dot, or keep the original string
                        s.strip_suffix(ext)
                            .and_then(|s| s.strip_suffix('.'))
                            .or(Some(s))
                    })
                    .unwrap_or("web_source")
            } else {
                "web_source"
            };

            let id = idr.resolve(id, url.to_string());
            configs.insert(id.clone(), FileConfigSrc::Path(path));
            results.push(cfg.custom.new_sources_url(id.clone(), url.clone()).await?);
            info!("Configured source {id} from URL {}", sanitize_url(&url));
        } else {
            let is_dir = path.is_dir();
            let dir_files = if is_dir {
                // directories will be kept in the config just in case there are new files
                directories.push(path.clone());
                collect_files_with_extension(&path, extension)?
            } else if path.is_file() {
                vec![path]
            } else {
                return Err(InvalidFilePath(path.canonicalize().unwrap_or(path)));
            };
            for path in dir_files {
                let can = path.canonicalize().map_err(|e| IoError(e, path.clone()))?;
                if files.contains(&can) {
                    if !is_dir {
                        warn!("Ignoring duplicate MBTiles path: {}", can.display());
                    }
                    continue;
                }
                let id = path.file_stem().map_or_else(
                    || "_unknown".to_string(),
                    |s| s.to_string_lossy().to_string(),
                );
                let id = idr.resolve(&id, can.to_string_lossy().to_string());
                info!("Configured source {id} from {}", can.display());
                files.insert(can);
                configs.insert(id.clone(), FileConfigSrc::Path(path.clone()));
                results.push(cfg.custom.new_sources(id, path).await?);
            }
        }
    }

    *config = FileConfigEnum::new_extended(directories, configs, cfg.custom);

    Ok(results)
}

/// Returns a vector of file paths matching any `allowed_extension` within the given directory.
///
/// # Errors
///
/// Returns an error if Rust's underlying [`read_dir`](std::fs::read_dir) returns an error.
fn collect_files_with_extension(
    base_path: &Path,
    allowed_extension: &[&str],
) -> Result<Vec<PathBuf>, FileError> {
    Ok(base_path
        .read_dir()
        .map_err(|e| IoError(e, base_path.to_path_buf()))?
        .filter_map(Result::ok)
        .filter(|f| {
            f.path()
                .extension()
                .filter(|actual_ext| {
                    allowed_extension
                        .iter()
                        .any(|expected_ext| expected_ext == actual_ext)
                })
                .is_some()
                && f.path().is_file()
        })
        .map(|f| f.path())
        .collect())
}

fn sanitize_url(url: &Url) -> String {
    let mut result = format!("{}://", url.scheme());
    if let Some(host) = url.host_str() {
        result.push_str(host);
    }
    if let Some(port) = url.port() {
        result.push(':');
        result.push_str(&port.to_string());
    }
    result.push_str(url.path());
    result
}

fn parse_url(is_enabled: bool, path: &Path) -> Result<Option<Url>, FileError> {
    if !is_enabled {
        return Ok(None);
    }
    path.to_str()
        .filter(|v| v.starts_with("http://") || v.starts_with("https://"))
        .map(|v| Url::parse(v).map_err(|e| InvalidSourceUrl(e, v.to_string())))
        .transpose()
}
