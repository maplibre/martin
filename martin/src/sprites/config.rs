use serde::{Deserialize, Serialize};

use crate::config::UnrecognizedValues;
use crate::file_config::ConfigExtras;

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct SpriteConfig {
    #[serde(flatten)]
    pub unrecognized: UnrecognizedValues,
}

impl ConfigExtras for SpriteConfig {
    fn get_unrecognized(&self) -> &UnrecognizedValues {
        &self.unrecognized
    }
}
