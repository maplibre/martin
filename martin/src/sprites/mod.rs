use std::collections::BTreeMap;
use std::fmt::Debug;
use std::path::PathBuf;

use dashmap::{DashMap, Entry};
use futures::future::try_join_all;
use log::{info, warn};
use serde::{Deserialize, Serialize};
use spreet::resvg::usvg::{Error as ResvgError, Options, Tree, TreeParsing};
use spreet::{
    SpreetError, Sprite, Spritesheet, SpritesheetBuilder, get_svg_input_paths, sprite_name,
};
use tokio::io::AsyncReadExt;

use self::SpriteError::{SpriteInstError, SpriteParsingError, SpriteProcessingError};
use crate::config::UnrecognizedValues;
use crate::file_config::{ConfigExtras, FileConfigEnum, FileResult};

pub type SpriteResult<T> = Result<T, SpriteError>;

#[derive(thiserror::Error, Debug)]
pub enum SpriteError {
    #[error("Sprite {0} not found")]
    SpriteNotFound(String),

    #[error("IO error {0}: {1}")]
    IoError(std::io::Error, PathBuf),

    #[error("Sprite path is not a file: {0}")]
    InvalidFilePath(PathBuf),

    #[error("Sprite {0} uses bad file {1}")]
    InvalidSpriteFilePath(String, PathBuf),

    #[error("No sprite files found in {0}")]
    NoSpriteFilesFound(PathBuf),

    #[error("Sprite {0} could not be loaded")]
    UnableToReadSprite(PathBuf),

    #[error("{0} in file {1}")]
    SpriteProcessingError(SpreetError, PathBuf),

    #[error("{0} in file {1}")]
    SpriteParsingError(ResvgError, PathBuf),

    #[error("Unable to generate spritesheet")]
    UnableToGenerateSpritesheet,

    #[error("Unable to create a sprite from file {0}")]
    SpriteInstError(PathBuf),
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct CatalogSpriteEntry {
    pub images: Vec<String>,
}

pub type SpriteCatalog = DashMap<String, CatalogSpriteEntry>;

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct SpriteConfig {
    #[serde(flatten)]
    pub unrecognized: UnrecognizedValues,
}

impl ConfigExtras for SpriteConfig {
    fn get_unrecognized(&self) -> &UnrecognizedValues {
        &self.unrecognized
    }
}

#[derive(Debug, Clone, Default)]
pub struct SpriteSources(DashMap<String, SpriteSource>);

impl SpriteSources {
    pub fn resolve(config: &mut FileConfigEnum<SpriteConfig>) -> FileResult<Self> {
        let Some(cfg) = config.extract_file_config(None)? else {
            return Ok(Self::default());
        };

        let mut results = Self::default();
        let mut directories = Vec::new();
        let mut configs = BTreeMap::new();

        if let Some(sources) = cfg.sources {
            for (id, source) in sources {
                configs.insert(id.clone(), source.clone());
                results.add_source(id, source.abs_path()?);
            }
        };

        for path in cfg.paths {
            let Some(name) = path.file_name() else {
                warn!(
                    "Ignoring sprite source with no name from {}",
                    path.display()
                );
                continue;
            };
            directories.push(path.clone());
            results.add_source(name.to_string_lossy().to_string(), path);
        }

        *config = FileConfigEnum::new_extended(directories, configs, cfg.custom);

        Ok(results)
    }

    pub fn get_catalog(&self) -> SpriteResult<SpriteCatalog> {
        // TODO: all sprite generation should be pre-cached
        let entries = SpriteCatalog::new();
        for source in &self.0 {
            let paths = get_svg_input_paths(&source.path, true)
                .map_err(|e| SpriteProcessingError(e, source.path.clone()))?;
            let mut images = Vec::with_capacity(paths.len());
            for path in paths {
                images.push(
                    sprite_name(&path, &source.path)
                        .map_err(|e| SpriteProcessingError(e, source.path.clone()))?,
                );
            }
            images.sort();
            entries.insert(source.key().clone(), CatalogSpriteEntry { images });
        }
        Ok(entries)
    }

    fn add_source(&mut self, id: String, path: PathBuf) {
        let disp_path = path.display();
        if path.is_file() {
            warn!("Ignoring non-directory sprite source {id} from {disp_path}");
        } else {
            match self.0.entry(id) {
                Entry::Occupied(v) => {
                    warn!(
                        "Ignoring duplicate sprite source {} from {disp_path} because it was already configured for {}",
                        v.key(),
                        v.get().path.display()
                    );
                }
                Entry::Vacant(v) => {
                    info!("Configured sprite source {} from {disp_path}", v.key());
                    v.insert(SpriteSource { path });
                }
            }
        }
    }

    /// Given a list of IDs in a format "id1,id2,id3", return a spritesheet with them all.
    /// `ids` may optionally end with "@2x" to request a high-DPI spritesheet.
    pub async fn get_sprites(&self, ids: &str, as_sdf: bool) -> SpriteResult<Spritesheet> {
        let (ids, dpi) = if let Some(ids) = ids.strip_suffix("@2x") {
            (ids, 2)
        } else {
            (ids, 1)
        };

        let sprite_ids = ids
            .split(',')
            .map(|id| self.get(id))
            .collect::<SpriteResult<Vec<_>>>()?;

        get_spritesheet(sprite_ids.iter(), dpi, as_sdf).await
    }

    fn get(&self, id: &str) -> SpriteResult<SpriteSource> {
        match self.0.get(id) {
            Some(v) => Ok(v.clone()),
            None => Err(SpriteError::SpriteNotFound(id.to_string())),
        }
    }
}

#[derive(Clone, Debug)]
pub struct SpriteSource {
    path: PathBuf,
}

async fn parse_sprite(
    name: String,
    path: PathBuf,
    pixel_ratio: u8,
    as_sdf: bool,
) -> SpriteResult<(String, Sprite)> {
    let on_err = |e| SpriteError::IoError(e, path.clone());

    let mut file = tokio::fs::File::open(&path).await.map_err(on_err)?;

    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).await.map_err(on_err)?;

    let tree = Tree::from_data(&buffer, &Options::default())
        .map_err(|e| SpriteParsingError(e, path.clone()))?;

    let sprite = if as_sdf {
        Sprite::new_sdf(tree, pixel_ratio)
    } else {
        Sprite::new(tree, pixel_ratio)
    };
    let sprite = sprite.ok_or_else(|| SpriteInstError(path.clone()))?;

    Ok((name, sprite))
}

pub async fn get_spritesheet(
    sources: impl Iterator<Item = &SpriteSource>,
    pixel_ratio: u8,
    as_sdf: bool,
) -> SpriteResult<Spritesheet> {
    // Asynchronously load all SVG files from the given sources
    let mut futures = Vec::new();
    for source in sources {
        let paths = get_svg_input_paths(&source.path, true)
            .map_err(|e| SpriteProcessingError(e, source.path.clone()))?;
        for path in paths {
            let name = sprite_name(&path, &source.path)
                .map_err(|e| SpriteProcessingError(e, source.path.clone()))?;
            futures.push(parse_sprite(name, path, pixel_ratio, as_sdf));
        }
    }
    let sprites = try_join_all(futures).await?;
    let mut builder = SpritesheetBuilder::new();
    if as_sdf {
        builder.make_sdf();
    }
    builder.sprites(sprites.into_iter().collect());

    // TODO: decide if this is needed and/or configurable
    // builder.make_unique();

    builder
        .generate()
        .ok_or(SpriteError::UnableToGenerateSpritesheet)
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::*;

    #[actix_rt::test]
    async fn test_sprites() {
        let mut cfg = FileConfigEnum::new(vec![
            PathBuf::from("../tests/fixtures/sprites/src1"),
            PathBuf::from("../tests/fixtures/sprites/src2"),
        ]);

        let sprites = SpriteSources::resolve(&mut cfg).unwrap().0;
        assert_eq!(sprites.len(), 2);

        //.sdf => generate sdf from png, add sdf == true
        //- => does not generate sdf, omits sdf == true
        for extension in ["_sdf", ""] {
            let paths = sprites
                .iter()
                .map(|v| v.value().clone())
                .collect::<Vec<_>>();
            test_src(paths.iter(), 1, "all_1", extension).await;
            test_src(paths.iter(), 2, "all_2", extension).await;

            let src1_path = sprites
                .get("src1")
                .into_iter()
                .map(|v| v.value().clone())
                .collect::<Vec<_>>();
            test_src(src1_path.iter(), 1, "src1_1", extension).await;
            test_src(src1_path.iter(), 2, "src1_2", extension).await;

            let src2_path = sprites
                .get("src2")
                .into_iter()
                .map(|v| v.value().clone())
                .collect::<Vec<_>>();
            test_src(src2_path.iter(), 1, "src2_1", extension).await;
            test_src(src2_path.iter(), 2, "src2_2", extension).await;
        }
    }

    async fn test_src(
        sources: impl Iterator<Item = &SpriteSource>,
        pixel_ratio: u8,
        filename: &str,
        extension: &str,
    ) {
        let path = PathBuf::from(format!(
            "../tests/fixtures/sprites/expected/{filename}{extension}"
        ));
        let sprites = get_spritesheet(sources, pixel_ratio, extension == "_sdf")
            .await
            .unwrap();
        let mut json = serde_json::to_string_pretty(sprites.get_index()).unwrap();
        json.push('\n');
        let png = sprites.encode_png().unwrap();

        #[cfg(feature = "bless-tests")]
        {
            use std::io::Write as _;
            let mut file = std::fs::File::create(path.with_extension("json")).unwrap();
            file.write_all(json.as_bytes()).unwrap();

            let mut file = std::fs::File::create(path.with_extension("png")).unwrap();
            file.write_all(&png).unwrap();
        }

        #[cfg(not(feature = "bless-tests"))]
        {
            let expected = std::fs::read_to_string(path.with_extension("json"))
                .expect("Unable to open expected JSON file, make sure to bless tests with\n  cargo test --features bless-tests\n");

            assert_eq!(
                serde_json::from_str::<serde_json::Value>(&json).unwrap(),
                serde_json::from_str::<serde_json::Value>(&expected).unwrap(),
                "Make sure to run bless if needed:\n  cargo test --features bless-tests\n\n{json}",
            );

            let expected = std::fs::read(path.with_extension("png"))
                .expect("Unable to open expected PNG file, make sure to bless tests with\n  cargo test --features bless-tests\n");

            // The PNG output is too flaky to be reliably used in a test
            if png != expected {
                warn!(
                    "Generated PNG does not match expected PNG, make sure to bless tests with\n  cargo test --features bless-tests\n"
                );
            }
            // assert_eq!(
            //     png, expected,
            //     "Make sure to run bless if needed:\n  cargo test --features bless-tests\n\n{json}",
            // );
        }
    }
}
