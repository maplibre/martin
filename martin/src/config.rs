use std::collections::HashMap;
use std::ffi::OsStr;
use std::fs::File;
use std::io::prelude::*;
use std::path::{Path, PathBuf};
use std::pin::Pin;

use futures::future::try_join_all;
use log::info;
use serde::{Deserialize, Serialize};
use subst::VariableMap;

use crate::MartinError::{ConfigLoadError, ConfigParseError, ConfigWriteError, NoSources};
#[cfg(any(feature = "fonts", feature = "postgres"))]
use crate::OptOneMany;
#[cfg(any(
    feature = "mbtiles",
    feature = "pmtiles",
    feature = "sprites",
    feature = "cog"
))]
use crate::file_config::FileConfigEnum;
#[cfg(feature = "fonts")]
use crate::fonts::FontSources;
use crate::source::{TileInfoSources, TileSources};
#[cfg(feature = "sprites")]
use crate::sprites::{SpriteConfig, SpriteSources};
use crate::srv::{RESERVED_KEYWORDS, SrvConfig};
use crate::utils::{CacheValue, MainCache, OptMainCache, init_aws_lc_tls, parse_base_path};
use crate::{IdResolver, MartinResult};

pub type UnrecognizedValues = HashMap<String, serde_yaml::Value>;

pub struct ServerState {
    pub cache: OptMainCache,
    pub tiles: TileSources,
    #[cfg(feature = "sprites")]
    pub sprites: SpriteSources,
    #[cfg(feature = "fonts")]
    pub fonts: FontSources,
}

#[serde_with::skip_serializing_none]
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct Config {
    pub cache_size_mb: Option<u64>,

    #[serde(flatten)]
    pub srv: SrvConfig,

    #[cfg(feature = "postgres")]
    #[serde(default, skip_serializing_if = "OptOneMany::is_none")]
    pub postgres: OptOneMany<crate::pg::PgConfig>,

    #[cfg(feature = "pmtiles")]
    #[serde(default, skip_serializing_if = "FileConfigEnum::is_none")]
    pub pmtiles: FileConfigEnum<crate::pmtiles::PmtConfig>,

    #[cfg(feature = "mbtiles")]
    #[serde(default, skip_serializing_if = "FileConfigEnum::is_none")]
    pub mbtiles: FileConfigEnum<crate::mbtiles::MbtConfig>,

    #[cfg(feature = "cog")]
    #[serde(default, skip_serializing_if = "FileConfigEnum::is_none")]
    pub cog: FileConfigEnum<crate::cog::CogConfig>,

    #[cfg(feature = "sprites")]
    #[serde(default, skip_serializing_if = "FileConfigEnum::is_none")]
    pub sprites: FileConfigEnum<SpriteConfig>,

    #[cfg(feature = "fonts")]
    #[serde(default, skip_serializing_if = "OptOneMany::is_none")]
    pub fonts: OptOneMany<PathBuf>,

    #[serde(flatten)]
    pub unrecognized: UnrecognizedValues,
}

impl Config {
    /// Apply defaults to the config, and validate if there is a connection string
    pub fn finalize(&mut self) -> MartinResult<UnrecognizedValues> {
        let mut res = UnrecognizedValues::new();
        copy_unrecognized_config(&mut res, "", &self.unrecognized);

        if let Some(path) = &self.srv.base_path {
            self.srv.base_path = Some(parse_base_path(path)?);
        }

        #[cfg(feature = "postgres")]
        for pg in self.postgres.iter_mut() {
            res.extend(pg.finalize()?);
        }

        #[cfg(feature = "pmtiles")]
        res.extend(self.pmtiles.finalize("pmtiles."));

        #[cfg(feature = "mbtiles")]
        res.extend(self.mbtiles.finalize("mbtiles."));

        #[cfg(feature = "cog")]
        res.extend(self.cog.finalize("cog."));

        #[cfg(feature = "sprites")]
        res.extend(self.sprites.finalize("sprites."));

        // TODO: support for unrecognized fonts?
        // res.extend(self.fonts.finalize("fonts.")?);

        let is_empty = true;

        #[cfg(feature = "postgres")]
        let is_empty = is_empty && self.postgres.is_empty();

        #[cfg(feature = "pmtiles")]
        let is_empty = is_empty && self.pmtiles.is_empty();

        #[cfg(feature = "mbtiles")]
        let is_empty = is_empty && self.mbtiles.is_empty();

        #[cfg(feature = "cog")]
        let is_empty = is_empty && self.cog.is_empty();

        #[cfg(feature = "sprites")]
        let is_empty = is_empty && self.sprites.is_empty();

        #[cfg(feature = "fonts")]
        let is_empty = is_empty && self.fonts.is_empty();

        if is_empty { Err(NoSources) } else { Ok(res) }
    }

    pub async fn resolve(&mut self) -> MartinResult<ServerState> {
        init_aws_lc_tls()?;
        let resolver = IdResolver::new(RESERVED_KEYWORDS);
        let cache_size = self.cache_size_mb.unwrap_or(512) * 1024 * 1024;
        let cache = if cache_size > 0 {
            info!("Initializing main cache with maximum size {cache_size}B");
            Some(
                MainCache::builder()
                    .weigher(|_key, value: &CacheValue| -> u32 {
                        match value {
                            CacheValue::Tile(v) => v.len().try_into().unwrap_or(u32::MAX),
                            #[cfg(feature = "pmtiles")]
                            CacheValue::PmtDirectory(v) => {
                                v.get_approx_byte_size().try_into().unwrap_or(u32::MAX)
                            }
                        }
                    })
                    .max_capacity(cache_size)
                    .build(),
            )
        } else {
            info!("Caching is disabled");
            None
        };

        Ok(ServerState {
            tiles: self.resolve_tile_sources(&resolver, cache.clone()).await?,
            #[cfg(feature = "sprites")]
            sprites: SpriteSources::resolve(&mut self.sprites)?,
            #[cfg(feature = "fonts")]
            fonts: FontSources::resolve(&mut self.fonts)?,
            cache,
        })
    }

    async fn resolve_tile_sources(
        &mut self,
        #[allow(unused_variables)] idr: &IdResolver,
        #[allow(unused_variables)] cache: OptMainCache,
    ) -> MartinResult<TileSources> {
        #[allow(unused_mut)]
        let mut sources: Vec<Pin<Box<dyn Future<Output = MartinResult<TileInfoSources>>>>> =
            Vec::new();

        #[cfg(feature = "postgres")]
        for s in self.postgres.iter_mut() {
            sources.push(Box::pin(s.resolve(idr.clone())));
        }

        #[cfg(feature = "pmtiles")]
        if !self.pmtiles.is_empty() {
            let cfg = &mut self.pmtiles;
            let val = crate::file_config::resolve_files(cfg, idr, cache.clone(), &["pmtiles"]);
            sources.push(Box::pin(val));
        }

        #[cfg(feature = "mbtiles")]
        if !self.mbtiles.is_empty() {
            let cfg = &mut self.mbtiles;
            let val = crate::file_config::resolve_files(cfg, idr, cache.clone(), &["mbtiles"]);
            sources.push(Box::pin(val));
        }

        #[cfg(feature = "cog")]
        if !self.cog.is_empty() {
            let cfg = &mut self.cog;
            let val = crate::file_config::resolve_files(cfg, idr, cache.clone(), &["tif", "tiff"]);
            sources.push(Box::pin(val));
        }

        Ok(TileSources::new(try_join_all(sources).await?))
    }

    pub fn save_to_file(&self, file_name: PathBuf) -> MartinResult<()> {
        let yaml = serde_yaml::to_string(&self).expect("Unable to serialize config");
        if file_name.as_os_str() == OsStr::new("-") {
            info!("Current system configuration:");
            println!("\n\n{yaml}\n");
            Ok(())
        } else {
            info!(
                "Saving config to {}, use --config to load it",
                file_name.display()
            );
            match File::create(&file_name) {
                Ok(mut file) => file
                    .write_all(yaml.as_bytes())
                    .map_err(|e| ConfigWriteError(e, file_name)),
                Err(e) => Err(ConfigWriteError(e, file_name)),
            }
        }
    }
}

pub fn copy_unrecognized_config(
    result: &mut UnrecognizedValues,
    prefix: &str,
    unrecognized: &UnrecognizedValues,
) {
    result.extend(
        unrecognized
            .iter()
            .map(|(k, v)| (format!("{prefix}{k}"), v.clone())),
    );
}

/// Read config from a file
pub fn read_config<'a, M>(file_name: &Path, env: &'a M) -> MartinResult<Config>
where
    M: VariableMap<'a>,
    M::Value: AsRef<str>,
{
    let mut file = File::open(file_name).map_err(|e| ConfigLoadError(e, file_name.into()))?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)
        .map_err(|e| ConfigLoadError(e, file_name.into()))?;
    parse_config(&contents, env, file_name)
}

pub fn parse_config<'a, M>(contents: &str, env: &'a M, file_name: &Path) -> MartinResult<Config>
where
    M: VariableMap<'a>,
    M::Value: AsRef<str>,
{
    subst::yaml::from_str(contents, env).map_err(|e| ConfigParseError(e, file_name.into()))
}

#[cfg(feature = "postgres")]
#[cfg(test)]
pub mod tests {
    use super::*;
    use crate::config::Config;
    use crate::tests::FauxEnv;

    pub fn parse_cfg(yaml: &str) -> Config {
        parse_config(yaml, &FauxEnv::default(), Path::new("<test>")).unwrap()
    }

    pub fn assert_config(yaml: &str, expected: &Config) {
        let mut config = parse_cfg(yaml);
        let res = config.finalize().unwrap();
        assert!(res.is_empty(), "unrecognized config: {res:?}");
        assert_eq!(&config, expected);
    }
}
