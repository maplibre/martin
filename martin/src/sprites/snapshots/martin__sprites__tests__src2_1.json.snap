---
source: martin/src/sprites/mod.rs
expression: sprites.get_index()
---
{
  "bicycle": {
    "height": 15,
    "pixelRatio": 1,
    "width": 15,
    "x": 0,
    "y": 0
  }
}
